### PR TITLE
plan, statistics: begin to use statsProfile to estimate count and cardinality.

### PR DESCRIPTION
**Pull request by hanfei1991**
This improvement will help estimate count and cardinality for every plan and get more precise result.

@winoros @lamxTyler @zimulala PTAL

### CODE REVIEW

File: statistics/selectivity.go [hunk 1]
```diff
@@ -56,8 +87,11 @@ func (t *Table) Selectivity(ctx context.Context, exprs []expression.Expression)
 	}
 	// TODO: If len(exprs) is bigger than 63, we could use bitset structure to replace the int64.
 	// This will simplify some code and speed up if we use this rather than a boolean slice.
-	if t.Pseudo || len(exprs) > 63 {
-		return selectionFactor, nil
+	if t.Pseudo || len(exprs) > 63 || len(t.Columns) == 0 {
```
Comment (lamxTyler) [inline statistics/selectivity.go]: Why only check the columns? What about the index?

File: plan/task.go [hunk 2]
```diff
@@ -197,18 +188,23 @@ func finishCopTask(task task, ctx context.Context, allocator *idAllocator) task
 	// `NetWorkStartCost` * (totalCount / perCountIndexRead)
```
Comment (winoros) [inline plan/task.go]: Is the `Start` in `NetWorkStartCost` a typo?

Comment (hanfei1991) [inline plan/task.go]: Nope

File: plan/stats.go [hunk 3]
```diff
@@ -26,6 +26,17 @@ type statsProfile struct {
 	cardinality []float64
 }
 
+func (s *statsProfile) collapse(selectivity float64) *statsProfile {
```
Comment (shenli) [inline plan/stats.go]: Add comment for this function.

File: plan/task.go [hunk 4]
```diff
@@ -179,11 +175,6 @@ func (p *PhysicalHashSemiJoin) attach2Task(tasks ...task) task {
 		// TODO: we will estimate the cost and count more precisely.
 		cst: lTask.cost() + rTask.cost(),
 	}
-	if p.WithAux {
```
Comment (shenli) [inline plan/task.go]: Why remove this?

File: statistics/selectivity.go [hunk 5]
```diff
@@ -45,6 +46,36 @@ const (
 	colType
 )
 
+func checkColumnConstant(e []expression.Expression) bool {
```
Comment (shenli) [inline statistics/selectivity.go]: Add comment

File: plan/task.go (hunk 4 above)
Comment (hanfei1991) [inline plan/task.go]: No need to calculate count

**Comment by lamxTyler** [comment]
LGTM

**Pull request merged**